drivers/spi/spi_mcux_flexcomm.c

```diff
@@ -218,6 +245,372 @@ static int spi_mcux_configure(const struct device *dev,
+}
+
+static int spi_mcux_dma_rx_load(const struct device *dev, uint8_t *buf,
+				 size_t len)
+{
+	const struct spi_mcux_config *cfg = dev->config;
+	struct spi_mcux_data *data = dev->data;
+	struct dma_block_config *blk_cfg;
+	int ret;
+	SPI_Type *base = cfg->base;
+
+	/* retrieve active RX DMA channel (used in callback) */
+	struct stream *stream = &data->dma_rx;
+
+	blk_cfg = &stream->dma_blk_cfg[0];
+
+	/* prepare the block for this RX DMA channel */
+	memset(blk_cfg, 0, sizeof(struct dma_block_config));
+	blk_cfg->block_size = len;
+
+	/* rx direction has periph as source and mem as dest. */
+	if (buf == NULL) {
+		/* if rx buff is null, then write data to dummy address. */
+		blk_cfg->dest_address = (uint32_t)&dummy_rx_buffer;
+	} else {
+		blk_cfg->dest_address = (uint32_t)buf;
+	}
+
+	blk_cfg->source_address = (uint32_t)&base->FIFORD;
+
+	/* direction is given by the DT */
+	stream->dma_cfg.head_block = blk_cfg;
+	stream->dma_cfg.user_data = data;
+
+	/* Enables the DMA request from SPI rxFIFO */
+	base->FIFOCFG |= SPI_FIFOCFG_DMARX_MASK;
+
+	/* pass our client origin to the dma: data->dma_rx.channel */
+	ret = dma_config(data->dma_rx.dma_dev, data->dma_rx.channel,
+			&stream->dma_cfg);
+	/* the channel is the actual stream from 0 */
+	if (ret != 0) {
+		return ret;
+	}
+
+	/* gives the request ID */
+	return dma_start(data->dma_rx.dma_dev, data->dma_rx.channel);
+}
+
+static int spi_mcux_dma_move_buffers(const struct device *dev, size_t len,
+			const struct spi_config *spi_cfg, bool last_packet)
+{
+	struct spi_mcux_data *data = dev->data;
+	int ret;
+
+	ret = spi_mcux_dma_rx_load(dev, data->ctx.rx_buf, len);
+
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = spi_mcux_dma_tx_load(dev, data->ctx.tx_buf, spi_cfg,
+							len, last_packet);
+
+	return ret;
+}
+
+static int wait_dma_rx_tx_done(const struct device *dev)
+{
+	struct spi_mcux_data *data = dev->data;
+	int ret = -1;
+
+	while (1) {
+		ret = spi_context_wait_for_completion(&data->ctx);
+		if (data->status_flags & SPI_MCUX_FLEXCOMM_DMA_ERROR_FLAG) {
+			return -EIO;
+		}
+
+		if ((data->status_flags & SPI_MCUX_FLEXCOMM_DMA_DONE_FLAG) ==
+			SPI_MCUX_FLEXCOMM_DMA_DONE_FLAG) {
+			return 0;
+		}
+	}
+}
+
+static int transceive_dma(const struct device *dev,
+		      const struct spi_config *spi_cfg,
+		      const struct spi_buf_set *tx_bufs,
+		      const struct spi_buf_set *rx_bufs,
+		      bool asynchronous,
+		      struct k_poll_signal *signal)
+{
+	const struct spi_mcux_config *config = dev->config;
+	struct spi_mcux_data *data = dev->data;
+	SPI_Type *base = config->base;
+	int ret;
+	uint32_t word_size;
+
+	spi_context_lock(&data->ctx, asynchronous, signal);
+
+	ret = spi_mcux_configure(dev, spi_cfg);
+	if (ret) {
+		goto out;
+	}
+
+	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
+
+	spi_context_cs_control(&data->ctx, true);
+
+	word_size = SPI_WORD_SIZE_GET(spi_cfg->operation);
+
+	data->dma_rx.dma_cfg.dest_data_size = (word_size > 8) ?
```

Comment:
it is better to add check for framesize, if exceed 2, then return error

Reply:
I am not clear on this comment, can you please provide more details.

Reply:
SPI wordsize is frame_size defined with
`
word_size = SPI_WORD_SIZE_GET(spi_cfg->operation);
`
so if you limited to 2 bytes, then  suggest to add some checking here

Reply:
This is the datawidth and not the transfer length

Reply:
yes, the datawidth here has some limitation, so I suggest to add a check here

Reply:
I see the datawidth is already checked inside spi_mcux_configure() function.

Reply:
if you mean, this is not the case, the data_size checked here is for DMA settings, but as the datawidth is to ensure the length of a spi frame. as lpc driver only support 1 or 2 bytes, but some platform support 3 and 4 bytes.
```
if (config->source_data_size != 4U &&
		config->source_data_size != 2U &&
		config->source_data_size != 1U) {
		LOG_ERR("Source unit size error, %d", config->source_data_size);
		return -EINVAL;
	}

	if (config->dest_data_size != 4U &&
		config->dest_data_size != 2U &&
		config->dest_data_size != 1U) {
		LOG_ERR("Dest unit size error, %d", config->dest_data_size);
		return -EINVAL;
	}
```

Reply:
right, so this can be ignored.